testnet/stacks-node/src/nakamoto_node/miner.rs

```diff
@@ -317,6 +318,8 @@ impl BlockMinerThread {
                         }
                     }
                 }
+                self.wait_min_time_between_blocks()?;
```

Comment:
Should we add this function call at the point where we've done some setup (ie gathered mempool transactions) before waiting? That would essentially add the minimum latency possible while still ensuring there is this gap. I'm thinking of something like moving this to line ~1129

Reply:
The timestamp in the block header currently gets set at the beginning of the block mining, so I guess we need to wait before starting to mine. We could do something like this, so we get to try to mine a block, then check if its timestamp is in range, and if not, we just try again. I think that maybe the best of both options.

```rust
                match self.mine_block(&stackerdbs) {
                    Ok(x) => {
                        if x.header.timestamp < self.get_min_block_time() {
                            info!("Block mined too quickly, will try again");
                            continue;
                        }
                        break Some(x)
                    }
```

where `get_min_block_time()` is a new method, similar to, but replacing your `wait_min_time_between_blocks`.

Reply:
KK will clean up :)

Reply:
Without a wait, this really spams the logs and actually breaks my test because It triggers a ton of event observer mined block alerts. Would we want to alert to a mined block that is not even issued to the signers because of its time stamp validation failing? I originally had added this retry within mine_block but it made the function recursive which I didn't love... :P otherwise I can change it to return Option(x) and if its none, reattempt. All of these seem so much uglier to me XD